core/storage/trie/polkadot_trie/polkadot_trie_impl.cpp

```diff
@@ -22,6 +22,179 @@ OUTCOME_CPP_DEFINE_CATEGORY(kagome::storage::trie, PolkadotTrieImpl::Error, e) {
   return "Unknown error";
 }
 
+namespace {
+  using namespace kagome::storage::trie;
+
+  uint32_t getCommonPrefixLength(const KeyNibbles &first,
+                                 const KeyNibbles &second) {
+    auto &&[it, _] =
+        std::mismatch(first.begin(), first.end(), second.begin(), second.end());
+    return it - first.begin();
+  }
+
+  /**
+   * Fixes node type, merge nodes after children removal
+   * 1. if node has no children left, but has value, then turn to leaf.
+   * 2. if node has no children and no value (can't really happen anywhere but
+   clearPrefix), then reset node.
+   * 3. if node has 1 child and is not a value, merge it with this child, update
+   its key, and copy children if any
+   */
+  outcome::result<void> handleDeletion(
+      PolkadotTrie::NodePtr &parent,
+      const PolkadotTrieImpl::ChildRetrieveFunctor &retrieveChild) {
+    auto branch = std::dynamic_pointer_cast<BranchNode>(parent);
+    auto bitmap = branch->childrenBitmap();
+    if (bitmap == 0) {
+      if (parent->value) {
+        // turn branch node left with no children to a leaf node
+        parent = std::make_shared<LeafNode>(parent->key_nibbles, parent->value);
+      } else {
+        // this case actual only for clearPrefix, unreal situation in deletion
+        parent.reset();
+      }
+    } else if (branch->childrenNum() == 1 && !parent->value) {
+      size_t idx = 0;
+      for (idx = 0; idx < 16; idx++) {
+        bitmap >>= 1u;
+        if (bitmap == 0) {
+          break;
+        }
+      }
```

Comment:
Would be actually nice to move this to a separate small function, to take this cognitive burden off this function. It's not too hard to understand, but still requires some processing.

Reply:
Accepted

Reply:
But I'll try to do it in next request, because I need these changes urgently